lib/internal/http2.js

```diff
@@ -0,0 +1,812 @@
+        // Consume as much of the currently buffered 
+        // data as possible per data frame up to buffer.length
+        return copyBuffers(buffer, chunks);
+      }
+    };
+
+    // If this Writable is connected to a pipe, resume any deferred data
+    // frames and initiate the response if it hasn't been initiated already.
+    this.on('pipe', () => {
+      this[kResume]();
+      this[kBeginResponse]();
+    });
+    
+    this.statusCode = 200;
+  }
+
+  get socket() {
+    return this[kSocket];
+  }
+
+  get finished() {
+    return this[kFinished];
+  }
+
+  get headersSent() {
+    return this[kHeadersSent];
+  }
+
+  get statusCode() {
+    this[kHeaders].get(':status');
+  }
+
+  set statusCode(code) {
+    code |= 0;
+    if (code < 100 || code > 999)
+      throw new RangeError(`Invalid status code: ${code}`);
+    this[kHeaders].set(':status', code);
+  }
+
+  setHeader(name, value) {
+    name = String(name).toLowerCase().trim();
+    // Delete the current value if it's null
+    if (value === undefined || value === null) {
+      this[kHeaders].delete(name);
+      return this;
+    }
+    // Cannot add headers that start with the :-prefix
+    if (name[0] === ':')
+      throw new TypeError('Cannot add HTTP/2 pseudo-headers');
+    if (Array.isArray(value)) {
+      this[kHeaders].set(name, value.map((i) => String(i)));
+    } else {
+      this[kHeaders].set(name, String(value));
+    }
+    return this;
+  }
+
+  setTrailer(name, value) {
+    name = String(name).toLowerCase().trim();
+    // Delete the current value if it's null
+    if (value === undefined || value === null) {
+      this[kTrailers].delete(name);
+      return this;
+    }
+    // Cannot add headers that start with the :-prefix
+    if (name[0] === ':')
+      throw new TypeError('Cannot add HTTP/2 pseudo-headers');
+    if (Array.isArray(value)) {
+      this[kTrailers].set(name, value.map((i) => String(i)));
+    } else {
+      this[kTrailers].set(name, String(value));
+    }
+    return this;
+  }
+
+  addTrailers(headers) {
+    var keys = Object.keys(headers);
+    for (var key of keys)
+      this.setTrailer(key, headers[key]);
+    return this;
+  }
+
+  getHeader(name) {
+    return this[kHeaders].get(name);
+  }
+
+  getTrailer(name) {
+    return this[kTrailers].get(name);
+  }
+
+  removeHeader(name) {
+    this[kHeaders].delete(name);
+    return this;
+  }
+
+  removeTrailer(name) {
+    this[kTrailers].delete(name);
+    return this;
+  }
+
+  setTimeout(msec, callback) {
+    if (callback)
+      this.on('timeout', callback);
+    this.socket.setTimeout(msecs);
+    return this;
+  }
+
+  writeContinue() {
+    this[kSession].continue(this[kStream]);
+  }
+
+  writeHead(statusCode, headers) {
+    this.statusCode = statusCode;
+    const keys = Object.keys(headers);
+    for (var key of keys)
+      this.setHeader(key, headers[key]);
+    return this;
+  }
+
+  _write(chunk, encoding, callback) {
+    if (typeof chunk === 'string')
```

Comment:
Do we need the encoding if a chunk is a Buffer?  Or, in other words, should encoding be an optional param, as in `end` directly below?